### ISSUE TITLE
`@AlsoNegation` for reference equality

### ISSUE THREAD

**Issue by Stephan202** [google/error-prone#559]
I defined the following Refaster template and ran it over our code base:
```java
public final class EnumEquals<T extends Enum<T>> {
    @BeforeTemplate
    boolean before(T a, T b) {
        return a.equals(b);
    }   

    @BeforeTemplate
    boolean before2(T a, T b) {
        return Objects.equals(a, b); 
    }   

    @AlsoNegation
    @AfterTemplate
    boolean after(T a, T b) {
        return a == b;
    }   
}
```

This resulted in expressions of the form `!(Enum.CONSTANT == someVar)`, while I would have expected `Enum.CONSTANT != someVar`. I fixed this using this obvious additional template:
```java
public final class EnumNotEquals<T extends Enum<T>> {
    @BeforeTemplate
    boolean before(T a, T b) {
        return !(a == b); 
    }   

    @AfterTemplate
    boolean after(T a, T b) {
        return a != b;
    }   
}
```

But my question is: can `@AlsoNegation` also be supported for reference equality? For expressions over primitive types it works really well :).

**Comment by eaftan** [comment]
@lowasser for Refaster expertise

**Comment by Stephan202** [comment]
It seems this has nothing to do with reference equality; stepping through the code with a debugger the produced `RefasterRule`s look fine. Rather, in this case the `@BeforeTemplate` expressions and their negations _both_ match, and the non-negated expressions seem to take precedence.

The following change "fixes" it:
```diff
diff --git a/core/src/main/java/com/google/errorprone/refaster/RefasterRuleBuilderScanner.java b/core/src/main/java/com/google/errorprone/refaster/RefasterRuleBuilderScanner.java
index 1b726d40e..cd197c1d6 100644
--- a/core/src/main/java/com/google/errorprone/refaster/RefasterRuleBuilderScanner.java
+++ b/core/src/main/java/com/google/errorprone/refaster/RefasterRuleBuilderScanner.java
@@ -186,7 +186,7 @@ public final class RefasterRuleBuilderScanner extends SimpleTreeVisitor<Void, Vo
         RefasterRule<?, ?> negation = RefasterRule.create(
             qualifiedTemplateClass, typeVars, negatedBeforeTemplates, 
             negatedAfterTemplates, annotationMap);
-        return ImmutableList.of(rule, negation);
+        return ImmutableList.of(negation, rule);
       }
       return ImmutableList.of(rule);
     }
```
(A more general approach may be to order expressions by the size of their AST, with larger expressions being tried first.)

**Comment by Stephan202** [comment]
For anybody reading along at home: a workaround that does not require a code change is to define the rule in terms of its negation, i.e.:

```java
final class EnumEquals<T extends Enum<T>> {
    @BeforeTemplate
    @SuppressWarnings("boxing")
    boolean before(T a, T b) {
        return !Refaster.anyOf(a.equals(b), Objects.equals(a, b));
    }

    @AlsoNegation
    @AfterTemplate
    boolean after(T a, T b) {
        return a != b;
    }
}
```

**Comment by Stephan202** [comment]
Hmm, afaik this issue is not resolved. IIUC it could be resolved by changing `RefasterScanner#scan` such that in case of conflicts the "largest" replacement is applied. We have some internal code that works around the issue by collecting the `Description`s emitted by Refaster and then applying a maximal subset using the following selection logic:
```java
/**
 * Reports a subset of the given matches, such that no two reported matches suggest a replacement
 * of the same part of the source code.
 *
 * <p>In the common case all matches will be reported. In case of overlap the match which replaces
 * the largest piece of source code is preferred. In case two matches wish to replace exactly the
 * same piece of code, preference is given to the match which suggests the shortest replacement.
 */
// XXX: This selection logic solves an issue described in
// https://github.com/google/error-prone/issues/559. Consider contributing it back upstream.
private static void applyMatches(
    Iterable<Description> allMatches, EndPosTable endPositions, VisitorState state) {
  ImmutableList<Description> byReplacementSize =
      ImmutableList.sortedCopyOf(
          Comparator.<Description>comparingInt(d -> getReplacedCodeSize(d, endPositions))
              .reversed()
              .thenComparingInt(d -> getInsertedCodeSize(d, endPositions)),
          allMatches);

  RangeSet<Integer> replacedSections = TreeRangeSet.create();
  for (Description description : byReplacementSize) {
    ImmutableRangeSet<Integer> ranges = getReplacementRanges(description, endPositions);
    if (ranges.asRanges().stream().noneMatch(replacedSections::intersects)) {
      /* This suggested fix does not overlap with any ("larger") replacement seen until now. Apply it. */
      state.reportMatch(description);
      replacedSections.addAll(ranges);
    }
  }
}

private static int getReplacedCodeSize(Description description, EndPosTable endPositions) {
  return getReplacements(description, endPositions).mapToInt(Replacement::length).sum();
}

// XXX: It might be nicer to prefer the shortest replacement _post formatting_.
private static int getInsertedCodeSize(Description description, EndPosTable endPositions) {
  return getReplacements(description, endPositions).mapToInt(r -> r.replaceWith().length()).sum();
}

private static ImmutableRangeSet<Integer> getReplacementRanges(
    Description description, EndPosTable endPositions) {
  return getReplacements(description, endPositions)
      .map(Replacement::range)
      .collect(toImmutableRangeSet());
}

private static Stream<Replacement> getReplacements(
    Description description, EndPosTable endPositions) {
  return description.fixes.stream().flatMap(fix -> fix.getReplacements(endPositions).stream());
}
```

@cushon or @lowasser would you be interested in a PR which integrates such logic directly in `RefasterScanner`? (Eye-balling it this shouldn't be too hard.)

**Comment by cushon** [comment]
From https://github.com/google/error-prone/issues/559#issuecomment-285967784 and https://github.com/google/error-prone/issues/559#issuecomment-285968658, I understood that it's testing the templates in order, and you can get the behavior you want by changing the order of the templates and the negation of the expression used with `@AlsoNegation`?

It's not obvious to me that a heuristic based on the replacement length would be the right choice for all templates.

**Comment by Stephan202** [comment]
> From [#559 (comment)](https://github.com/google/error-prone/issues/559#issuecomment-285967784) and [#559 (comment)](https://github.com/google/error-prone/issues/559#issuecomment-285968658), I understood that it's testing the templates in order, and you can get the behavior you want by changing the order of the templates and the negation of the expression used with `@AlsoNegation`?

Using the negation of the expression works indeed, but from the user's perspective (well, at least one user's perspective :wink: ) it isn't obvious; I only found out after debugging the Refaster code.

> It's not obvious to me that a heuristic based on the replacement length would be the right choice for all templates.

That's a fair concern. Anecdotally I can tell that for the collection of ~675 Refaster templates we have internally this heuristic works very well. It's especially helpful because we apply many templates in bulk (loaded from the classpath), and they don't have an explicit ordering between them. (Nor is such an ordering something we'd want to maintain at scale. :sweat_smile:)

That said, if a "transparent" change such as suggested above isn't deemed suitable, perhaps we could document the behavior on the `@AlsoNegation` annotation? Something like:

> Note that the explicitly defined rule is applied before its negation. If the latter structurally embeds the former (e.g. using a {@code !} prefix), consider defining the Refaster template in terms of the negation.

But now that I wrote this down: it seems that it's precisely in the case where Refaster adds (or removes) a `!` prefix that this issue arises. So another approach is that we identify this case in `RefasterRuleBuilderScanner` and order the two created rules accordingly. If possible, that would make for a very localized change. (But looking at `UExpression#negate()` it's not obvious to me how one would identify this case :thinking:.)
